### DISCUSSION TITLE
Lookup of dependencies from a private Gitlab Maven repository

### DISCUSSION THREAD

**Discussion by otbe** [renovatebot/renovate#8499] [Request Help]
Hi,

Im just wondering if there's currently a way to fetch dependency updates from private gitlab repository via Maven.
Depending on the token type you use you have to use different names for the HTTP header that you have to send alongside your requests. See https://docs.gitlab.com/ee/user/packages/maven_repository/#authenticate-to-the-package-registry-with-maven

I found an issue in the old renovate-config repo about this topic https://github.com/renovatebot/config-help/issues/998 and it indicates that it works (or worked) but I cant get it working CI tokens nor PAT.

In my renovate config I use something like:
```
    {
      "baseUrl": "https://gitlab.com/api/v4/groups/XXXXXXX/-/packages/maven",
      "encrypted": {
        "token": "<encryptedToken>"
      }
    }
```

and in my pom.xml something like:
```
	<repositories>
		<repository>
			<id>gitlab-maven</id>
			<url>https://gitlab.com/api/v4/groups/XXXXXXX/-/packages/maven</url>
		</repository>
	</repositories>
```

Any ideas what I am doing wrong here?

**Comment by rarkins** [comment]
How are you running Renovate? e.g. self-hosted or hosted app?

**Reply by otbe** [reply]
Currently we have a mixed approach. Some projects are on self-hosted and some on hosted.
We already use Gitlabs docker registry for our internal images which works quite fine

**Comment by rarkins** [comment]
What does the debug log "packageFiles with updates" look like right now for that gitlab-hosted package?

**Reply by otbe** [reply]
It contains the failed dependency with the proper registry url but also a warning
```
"Failed to look up dependency our.group.andpackage"
```

The output from the actually access says
```
DEBUG: Content is not found for Maven url: https://gitlab.com/api/v4/groups/XXXXXXX/-/packages/maven/our/group/andpackage/maven-metadata.xml (repository=affectedrepo)
```
If I visit the URL while being logged in I get the XML file but without login it will yield a 404

**Reply by otbe** [reply]
BTW I currently use the same token (PAT) for accessing the maven repository like for the Gitlab API in general. Its an dedicated CI user that has all the necessary permissions. That should be send as header `Private-Token` I guess from Gitlabs docs.

**Reply by rarkins** [reply]
The log I asked for should look something like this:

```
DEBUG: packageFiles with updates (repository=renovate-tests/pipelines)
       "config": {
         "bazel": [
           {
             "packageFile": "WORKSPACE",
             "deps": [
               {
                 "depType": "http_archive",
                 "managerData": {
                   "def": "http_archive(\n    name = \"bazel_toolchains\",\n    sha256 = \"4b1468b254a572dbe134cc1fd7c6eab1618a72acd339749ea343bd8f55c3b7eb\",\n    strip_prefix = \"bazel-toolchains-d665ccfa3e9c90fa789671bf4ef5f7c19c5715c4\",\n    urls = [\n        \"https://mirror.bazel.build/github.com/bazelbuild/bazel-toolchains/archive/d665ccfa3e9c90fa789671bf4ef5f7c19c5715c4.tar.gz\",\n        \"https://github.com/bazelbuild/bazel-toolchains/archive/d665ccfa3e9c90fa789671bf4ef5f7c19c5715c4.tar.gz\",\n    ],\n)"
                 },
                 "depName": "bazel_toolchains",
                 "repo": "bazelbuild/bazel-toolchains",
                 "currentDigest": "d665ccfa3e9c90fa789671bf4ef5f7c19c5715c4",
                 "datasource": "github-tags",
                 "lookupName": "bazelbuild/bazel-toolchains",
                 "depIndex": 0,
                 "updates": [
                   {
                     "updateType": "digest",
                     "newDigest": "471da0273050b88d77529484ff89741ff586f9f5",
                     "newDigestShort": "471da02"
                   }
                 ],
                 "warnings": []
               },
```

**Reply by otbe** [reply]
Yeah its just too huge to put it here :)
I stripped it down to the relevant parts
```
{
  "config": {
    "dockerfile": [
      {
        "packageFile": "Dockerfile",
        "deps": [
          ....
        ]
      }
    ],
    "maven": [
      {
        "datasource": "maven",
        "packageFile": "pom.xml",
        "deps": [
          {
            "datasource": "maven",
            "depName": "our.org.andPackage",
            "currentValue": "0.1.0",
            "fileReplacePosition": 4699,
            "registryUrls": [
              "https://repo.maven.apache.org/maven2",
              "https://gitlab.com/api/v4/groups/XXXXXX/-/packages/maven"
            ],
            "depIndex": 5,
            "updates": [],
            "warnings": [
              {
                "depName": "our.org.andPackage",
                "message": "Failed to look up dependency our.org.andPackage"
              }
            ],
            "fixedVersion": "0.1.0"
          }
        ],
        "parent": "../pom.xml"
      }
    ]
  }
}


```

**Reply by rarkins** [reply]
> That should be send as header `Private-Token` I guess from Gitlabs docs.

Where is that in the docs? I'm surprised if GitLab still use non-standard headers for anything. 

Try adding a package rule like this:

```
    {
      "baseUrl": "https://gitlab.com/api/v4/",
      "encrypted": {
        "token": "<encryptedToken>"
      }
    }
```

i.e. same token as you use for git - assuming you've verified it manually that it can access the packages too. It should then be passed in the `authorization` header.

**Reply by otbe** [reply]
I take this knowledge from their docs https://docs.gitlab.com/ee/user/packages/maven_repository/#authenticate-to-the-package-registry-with-maven

It doesn't work without the "custom" header field, even locally.

**Reply by rarkins** [reply]
Thanks, that would definitely explain it. @viceice I'm surprised that GitLab didn't add support for the standard header for maven at the same time they did for their API

**Reply by viceice** [reply]
🤷‍♂️

**Reply by rarkins** [reply]
Here's where you can see that PAT are supported by the API using standard headers: https://docs.gitlab.com/ee/api/

Is it a different access control mechanism for the maven registry?

**Reply by otbe** [reply]
It seems to work! At least locally I was able to use `Authorization: Bearer <token>`. Ill dive deeper into the issue later. Thanks!

**Reply by Burtan** [reply]
What is your final configuration?
